### PR TITLE
QBO freeze branch

### PR DESCRIPTION
**Pull request by hayata-suenaga**
| ID      | TITLE                                                                 | BRANCH                                                               | CREATED AT       |
|---------|-----------------------------------------------------------------------|----------------------------------------------------------------------|------------------|
| #42561  | SelectionList: Added listEmptyContent prop + QBO: Added an illustr... | s77rt:SelectionList-listEmptyContent                                 | about 7 days ago |
| #42548  | QBO: Add explicit description for each export entity                  | s77rt:qbo-export-out-of-pocket-expense-copy                          | about 8 days ago |
| #42315  | Enable / Disable multi level tags                                     | hayata-call-api-for-multi-level-tags                                 | about 15 days ago|
| #42149  | [NoQA] Add orderWeight to tag view and tag edit routes                | s77rt:multi-level-tags-fix-not-found                                 | about 17 days ago|
| #42004  | Prevent the user from requiring tags when there is no active tag      | hayata-do-not-allow-users-from-requiring-tags-when-there-is-no-ac... | about 21 days ago|
| #41638  | Prevent invalid configuration between location and JE selection       | hayata-do-not-allow-invalid-location-account-combo                   | about 27 days ago|
| #41310  | Last synced just now does not update                                  | tienifr:fix/41045                                                    | about 1 month ago|

### Details

Same logic as https://github.com/Expensify/App/pull/42340, https://github.com/Expensify/App/pull/42043
### Fixed Issues

$
Related to https://github.com/Expensify/Expensify/issues/201614

Since it's a feature branch that was created during the merge freeze, no specific issue had to be created.
Every PR merged here has been individually reviewed both by an internal engineer and a C+. Also every PR was linked to an issue.

### Tests

- [ ] Verify that no errors appear in the JS console
NA - individual PRs have the corresponding tests

### Offline tests

NA

### QA Steps

Please test all the steps listed below 🙇 

## #42561

Preconditions:
- Workspace connected to QBO

1. Go to Workspace > Accounting
2. Click on Export > Export invoices to
3. Verify that if there is no account to select then a message "No accounts found" is displayed
4. Click on Export > Export out-of-pocket expenses as > Account
5. Verify that if there is no account to select then a message "No accounts found" is displayed
6. Click on Export > Export company cards as > Account
7. Verify that if there is no account to select then a message "No accounts found" is displayed
8. Click on Export > Export company cards as > Vendor (You will need to change the entity to Vendor bill to see this option)
9. Verify that if there is no account to select then a message "No accounts found" is displayed
10. Click on Advanced > QuickBooks bill payment account
11. Verify that if there is no account to select then a message "No accounts found" is displayed
12. Click on Advanced > QuickBooks invoice collections account
13. Verify that if there is no account to select then a message "No accounts found" is displayed

## #42548

Preconditions:
- Workspace connected to QBO

1. Go to Workspace > Accounting
2. Click Export > Export out-of-pocket expenses as
3. Verify that changing the export entity changes the label of the account option as follow:

- Check -> Bank account
- Journal entry -> Account
- Vendor bill -> Accounts payable

4. Click on the account option for each of the above entities
5. Verify that the title of the page you are navigated to matches the label

## #42315

- Enable tags on a workspace
- Create several tags. Confirm that bulk enable and disable works by checking the top checkbox and clicking the Selected button.
- Enable the accounting feature and connect to QBO.
- Go back to the Tags screen and check that Classes and Customers/Projects are now imported as tags
- Click Classes. Click the top checkbox. Click the Selected button and click Disable
- Check that all tags in Classes are disabled
- Do the same thing but enable tags this time. Check that all tags in Classes are enabled back.
- Click Customers/Projects. Do the same thing as above. Confirm that they work.

## #42149

Preconditions:
- Accounting beta enabled
- Workspace connected to QBO

1. Go to workspace accounting and import Classes and Customers/Projects
2. Go to workspace tags
3. Click Classes then click on any subtag
4. Verify that the RHP for the selected subtag settings is opened
5. Go back to workspace tags
6. Click Customers/Projects then click on any subtag
7. Verify that the RHP for the selected subtag settings is opened

## #42004

1. Create a workspace
2. Enable the tags feature on the More Features page of the workspace editor
3. Create several tags manually on the Tags page
4. Click the Settings button on the Tags page and confirm that you can toggle on and off the Members must tag all spend toggle button
5. Disable all tags
6. Click the Settings button on the Tags page and confirm that the Members must tag all spend toggle button is off and locked.
7. Go back to More Features page and enable the Accounting feature
8. On the Accounting page, connect to QBO.
9. Go to the Tags page again
10. Check that classes and customers/projects are imported from QBO as tags and there are enabled
11. Click the Settings button on the Tags page and confirm that you can toggle on and off the Members must tag all spend toggle button
12. Disable all tags
13. Click the Settings button on the Tags page and confirm that the Members must tag all spend toggle button is off and locked.

## #41638

1. Create a new workspace
2. Go to the More Feature page and enable the accounting feature
3. Connect to QBO
4. Go to the Export page. Check that `Vendor Bill` is selected for `Export out-of-pocket expenses as` by default
5. Go to the Import page. Check the Location configuration. Check that the Switch is toggled off and locked. There should be a hint text under the Switch button. Use the screenshot below as a reference.
Screenshot 2024-05-10 at 5 44 21 PM

6. Go to the Export page. Select `Journal Entry` for `Export out-of-pocket expenses as`
7. Go to the Import page. Check Location. Confirm that now you can toggle the switch on or off (i.e. not locked)
8. Enable the Location import (i.e. toggle it on)
9. Export -> `Export out-of-pocket expenses as`
10. Confirm that Journal Entry is the only option displayed.
11. Check that there is a hint text as shown below

Screenshot 2024-05-29 at 4 25 19 PM

13. Go back and click the company card export menu
14. Confirm that the only two options displayed is Credit card or debit card
15. Check that there is a hint text displayed and that it looks like the one below

Screenshot 2024-05-29 at 4 27 25 PM

## #41310

1. Connect to QBO
2. Observe the "Last synced just now" text
3. Wait for several mins
4. Navigate away from the page and back or refresh page
5. Verify that the text changed correctly

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by hayata-suenaga** [comment]
@getusha, I'm assigned you since you were assigned to [this PR](https://github.com/Expensify/App/pull/42149)

**Comment by dubielzyk-expensify** [comment]
Is there anything to review here? If so, let us know. If not feel free to remove the Design as reviewer :)

**Comment by hayata-suenaga** [comment]
This is the branch to which we're merging all QBO related PRs.

We can test this and maybe potentially get the design review if necessary (but I think all PRs merged here were already reviewed.)

**Comment by hayata-suenaga** [comment]
Asking @getusha to solve the merge conflict of this PR

File: src/pages/workspace/tags/WorkspaceTagsSettingsPage.tsx [hunk 1]
```diff
@@ -66,6 +67,7 @@ function WorkspaceTagsSettingsPage({route, policyTags}: WorkspaceTagsSettingsPag
                                         isOn={policy?.requiresTag ?? false}
                                         accessibilityLabel={translate('workspace.tags.requiresTag')}
                                         onToggle={updateWorkspaceRequiresTag}
+                                        disabled={!policy?.areTagsEnabled || !hasEnabledOptions}
```
Comment (s77rt) [inline src/pages/workspace/tags/WorkspaceTagsSettingsPage.tsx:70]: If I enable "Members must tag all expenses" then disable all the tags, then the "Members must tag all expenses" switch is locked at ON.


https://github.com/Expensify/App/assets/16493223/db76ba22-2938-41de-958e-cc24fa642770


Solution: If tagging is required, the user shouldn't be able to disable the last enabled tag

Comment (s77rt) [reply src/pages/workspace/tags/WorkspaceTagsSettingsPage.tsx:70]: Could be done in a follow up

**Comment by s77rt** [comment]
I also noticed that offline first behaviour is broken, it's a regression from https://github.com/Expensify/App/pull/41627.

https://github.com/Expensify/App/assets/16493223/00a44bb6-0881-4f6f-911c-192e9bc9e842

**Comment by s77rt** [comment]
## Reviewer Checklist

- [X] I have verified the author checklist is complete (all boxes are checked off).
- [X] I verified the correct issue is linked in the `### Fixed Issues` section above
- [X] I verified testing steps are clear and they cover the changes made in this PR
    - [X] I verified the steps for local testing are in the `Tests` section
    - [X] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [X] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [X] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I verified tests pass on **all platforms** & I tested again on:
    - [X] Android: Native
    - [X] Android: mWeb Chrome
    - [X] iOS: Native
    - [X] iOS: mWeb Safari
    - [X] MacOS: Chrome / Safari
    - [X] MacOS: Desktop
- [X] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [X] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [X] I verified that all the inputs inside a form are aligned with each other.
    - [X] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by s77rt** [comment]
Besides the lint issue, code looks good to me and tests well

File: src/pages/workspace/tags/WorkspaceTagsSettingsPage.tsx (hunk 1 above)
Comment (trjExpensify) [reply src/pages/workspace/tags/WorkspaceTagsSettingsPage.tsx:70]: Maybe even here? https://github.com/Expensify/App/pull/42328

Comment (hayata-suenaga) [reply src/pages/workspace/tags/WorkspaceTagsSettingsPage.tsx:70]: Yep I think https://github.com/Expensify/App/pull/42328 is going to solve the issue. let me double check on that

**Comment by trjExpensify** [comment]
> mWeb

I'm going to remove this because all of the individual PRs were approved and reviewed, we just merged them here to be able to test in aggregate during the freeze. 👍

**Review by s77rt — APPROVED** [review]

**Review by aldo-expensify — APPROVED** [review]

**Pull request merged**